### PR TITLE
Render markdown comments with Marked and DOMPurify

### PR DESCRIPTION
**Pull request by ktuite**
Frontend renders comments with markdown in them (on the backend, they are saved as plain text/markdown).

This is using [Marked](https://www.npmjs.com/package/marked) to do the rendering and [DOMPurify](https://www.npmjs.com/package/dompurify) to do the sanitizing. 

Screen Shot 2021-04-27 at 4 48 40 PM

One small note, I took out the whitespace CSS because it was adding too much whitespace between lines, especially in the case of lists. I altered the `marked` options to add back in `<br>` line breaks so if a user does type new lines, they don't get pushed all onto a single line.

### CODE REVIEW

**Comment by issa-tseng** [comment]
fyi marked is ~43KB packed and dompurify is ~18KB packed.

**Comment by ktuite** [comment]
> fyi marked is ~43KB packed and dompurify is ~18KB packed.

How does that compare to the other packages?

**Comment by ktuite** [comment]
Fixed up the code to fix the tests, and fixed up the CSS so the padding is even.

Screen Shot 2021-04-28 at 11 28 32 AM

**Comment by issa-tseng** [comment]
> > fyi marked is ~43KB packed and dompurify is ~18KB packed.
> 
> How does that compare to the other packages?

not sure! matt would know. hence why just fyi :)

**Comment by matthew-white** [comment]
43 + 18 KB is definitely sizable. But since we lazy-load routes, it'll only load once the user navigates to the submission detail page, so it seems OK to me.

**Comment by issa-tseng** [comment]
okay! just bear in mind comments will end up in more places. and possibly a _lot_ of places.

File: src/components/submission/feed-entry.vue [hunk 1]
```diff
@@ -71,7 +74,10 @@ export default {
         : 'edited';
     },
     comment() {
-      return this.entry.notes != null ? this.entry.notes : this.entry.body;
+      const comment = this.entry.notes != null ? this.entry.notes : this.entry.body;
+      return comment != null
+        ? DOMPurify.sanitize(marked(comment, { gfm: true, breaks: true }))
```
Comment (matthew-white) [inline src/components/submission/feed-entry.vue:79]: Is there a way to add attributes to links? We usually specify `target="_blank"` for links that leave Central. And maybe it'd be good to specify `rel="noopener"` or something along those lines.

**Comment by matthew-white** [comment]
If there's a Markdown package that's smaller or that doesn't require an additional package to sanitize, that would be nice. But if these packages are used on multiple pages, webpack should put them in their own .js file, which the browser should cache.

**Review by matthew-white — COMMENTED** [review]
This is looking good!

Since a separate package is used to sanitize, how about we add a test of the sanitizing? Also, I think the release criteria mention adding a link to Markdown help.

**Comment by ktuite** [comment]
I'm also stripping out raw HTML from comments, using markdown-it (https://github.com/markdown-it/markdown-it, which seems to be larger than marked)

Screen Shot 2021-04-30 at 11 50 08 AM

**Comment by issa-tseng** [comment]
can we get that Markdown support to be dotted rather than dashed?

File: src/components/submission/feed-entry.vue [hunk 2]
```diff
@@ -71,9 +74,20 @@ export default {
...
+        : comment;
     }
   },
+  created() {
+    DOMPurify.addHook('afterSanitizeAttributes', (node) => {
```
[the remark this answers is not in the document]
Comment (ktuite) [reply src/components/submission/feed-entry.vue]: ah yes, done!

File: src/components/submission/feed-entry.vue [hunk 3]
```diff
@@ -71,9 +74,20 @@ export default {
...
+      if ('target' in node) {
```
[the remark this answers is not in the document]
Comment (ktuite) [reply src/components/submission/feed-entry.vue]: This is from the documentation: https://github.com/cure53/DOMPurify/blob/main/demos/README.md#hook-to-open-all-links-in-a-new-window-link 

Even `<a>` tags and markdown links that don't have a target explicitly set seem to get handled properly here... maybe node is a DOMPurify object that has this specified? I'm not sure how it's working under the hood but I believe it does work!

File: src/components/submission/comment.vue [hunk 4]
```diff
@@ -22,6 +22,7 @@ except according to the terms contained in the LICENSE file.
     </div>
     <div v-show="body !== '' || awaitingResponse || editWithoutComment"
       id="submission-comment-actions">
+      <a href="https://commonmark.org/help/" class="externalHelpLink" target="_blank">{{ $t('markdownSupported') }} </a>
```
[the remark this answers is not in the document]
Comment (ktuite) [reply src/components/submission/comment.vue]: I can add it here just in case someone doesn't have a modern browser...

**Review by matthew-white — APPROVED** [review]
Nice work, @ktuite! I think we're taking the right approach with marked + DOMPurify. I also think we'll be well positioned to extend this work when we add Markdown in more places and add Markdown preview.

I pushed a commit to disable an ESLint warning (not an error, which is why CircleCI didn't make an indication of it).

I initially pushed a commit to remove non-HTML elements from the Markdown. However, I ended up having larger questions around which tags and attributes to allow, so I removed that commit. I'll add a comment below about some of my questions there; I think any changes can happen in a follow-up PR.

I think we may also want to update the hover/focus styles for the link to Markdown help:

Hover
Screen Shot 2021-05-01 at 12 26 37 PM

Focus
Screen Shot 2021-05-01 at 12 26 29 PM

If so, that could also be done in a follow-up PR.

**Comment by matthew-white** [comment]
I like that we allow HTML in the Markdown. Both GitHub and the ODK Forum do as well, and I've used HTML on both sites in rare cases.

That said, I think by default DOMPurify will allow tags and attributes that aren't applicable in the context of Markdown. I think we should consider excluding:

- Non-HTML elements (SVG and MathML)
- `<style>` elements and `style` attributes
- Form controls and buttons
- `data` attributes that are used in Bootstrap's JavaScript (maybe the most important in this list)
- `id` and `class` attributes

Or rather than excluding specific tags and attributes, I think DOMPurify also makes it easy to specify just the ones we want. As a first pass, I think we could consider allowing these tags:

- a
- b
- br
- code
- em
- h1
- h2
- h3
- h4
- h5
- h6
- hr
- i
- img
- li
- ol
- p
- pre
- s
- small
- sub
- sup
- strong
- u
- ul

For attributes, I think we should allow `href`, `src`, and `alt`, and maybe also `width` and `height`.

**Pull request merged**